### PR TITLE
Update to sophia 3.2, fix minor bug

### PR DESCRIPTION
**Pull request by nikita-fuchs**
Just practicing. :)

### CODE REVIEW

**Comment by nikita-fuchs** [comment]
cc @maptuhec :)

**Comment by maptuhec** [comment]
@nikita-fuchs you just need to run `npm i -g forgae` and then in this project `forgae init --update` that will update to forgae's latest version and the changes will work

**Comment by AleksandarFilipov** [comment]
hm, i'm not able to compile this :(  `{:error, "Code errors\nlast_declaration_must_be_contract"}`.

**Comment by mradkov** [comment]
> hm, i'm not able to compile this :( `{:error, "Code errors\nlast_declaration_must_be_contract"}`.

It compiles just fine (just tried at https://testnet.contracts.aepps.com) with the resulting bytecode:

```
cb_+Q+bRgKgTYcsCVF2LAK8HQrGiL3k1C2ssHhHj39465wJBF/4zpP5CT75ATGgMQaxC83e9v4r4ZUPo/v+oEQZ7XfqzKv3GagYpzaq9tqLbmFtZV9leGlzdHO4wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACg//////////////////////////////////////////8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAbhAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPkBKqBoSzkSmV+fosrw4XFnypgv9zNWw7FxogQ0jCUnqJFCpIR0ZXN0uMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAoP//////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAG4QAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD5ATWghsI+qYmfa+LV/K8dwOny8yQ6+EJMy7nmDjCKj/YtRqWPZ2V0X2hhbXN0ZXJfZG5huMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAoP//////////////////////////////////////////AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAG4QAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD5AVSgobF16tgsbuH9EojyZdiVRE+HBPv/YjIzDKknbSa7ZoaOY3JlYXRlX2hhbXN0ZXK4wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACg//////////////////////////////////////////8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAbhgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA///////////////////////////////////////////+QRLoNYgYJqlfXV5OSn5Zp72ApZrbYTkQrIhQUlmbFgd0en/hGluaXS4YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAP//////////////////////////////////////////7kDwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEA//////////////////////////////////////////8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAeD//////////////////////////////////////////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAkAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAALgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAANAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA2AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOg//////////////////////////////////////////8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAALkGLmIAAOViAAGLkYCAgFF/obF16tgsbuH9EojyZdiVRE+HBPv/YjIzDKknbSa7ZoYUYgAB41dQgIBRf4bCPqmJn2vi1fyvHcDp8vMkOvhCTMu55g4wio/2LUalFGIABExXUICAUX/WIGCapX11eTkp+Wae9gKWa22E5EKyIUFJZmxYHdHp/xRiAATTV1CAgFF/MQaxC83e9v4r4ZUPo/v+oEQZ7XfqzKv3GagYpzaq9toUYgAGG1dQgFF/aEs5Eplfn6LK8OFxZ8qYL/czVsOxcaIENIwlJ6iRQqQUYgAGA1dQYAEZUQBbYAAZWWAgAZCBUmAgkANgABlZYCABkIFSYCCQA2AAWZCBUoFSWWAgAZCBUmAgkANgAVmQgVKBUllgIAGQgVJgIJADYABZkIFSgVJZYCABkIFSYCCQA2ADgVJZYEABkIFSYCCQA2ABWZCBUoFSYCCQA2AGgVKBUllgIAGQgVJgIJADYABZkIFSgVJZYCABkIFSYCCQA2ADgVKQWWAAUVlSYABSYADzW2AAgFJgAPNbYAD9kFCQVltQgGIAAzuQYABRUUJiAAORYAFDA2IAA5iQgEBgAIEUYgAB1VeAWWAgAZCBUmAgkANgAYFSYgAB3FZbYABZkIFSW5BQkFCQVltgIAFRUZBQWVCAkVBQf05hbWUgaXMgYWxyZWFkeSB0YWtlbgAAAAAAAAAAAAAAWWAgAZCBUmAgkANgFYFSgWIAArOQW2AAUWAgAVGBkJFQgIJiAAJjkWAAYABgAINZYEABkIFSYCCQA4WBUmAgkANhAS2BUmAAYABa8ZFQUJBWW4BRYAAUYgACd1dQYAFbkVBQkFZbUGAAYgACcVZbUGAAYABgAINZYGABkIFSYCCQA4WBUmAgkAOGgVJgIJADYQEugVJgAGAAWvGSUFBQkFZbFWIAAZuRW4FiAALKV4CRUFBiAAGTVltZkVBQkFZbYCABUYCQUICQUJBWW1CAYABRYCABUWIABMeRgIJiAAMdkWAAYABgAINZYEABkIFSYCCQA4WBUmAgkANhAS2BUmAAYABa8ZFQUJBWW4BRYAEUYgADLldQYAEZUQBbYCABUYCQUJFQUJBWW4GBkVCRUIBZYEABkIFSYCCQA4KBUmAgkANgAFFRgVJgAFGAYCABUWIABAyQg1lgYAGQgVJgIJADhoFSYCCQA1liAAKAgVKBUmAgkANiAAPqgVKAUVZbAwGQUJBWW4BRYAAUYgADs1eAUWABFGIAAtFXUGABGVEAW1B/YmxvY2toYXNoIG5vdCBmb3VuZAAAAAAAAAAAAAAAAABZYCABkIFSYCCQA2ATgVJiAAGTVluAUGAgAYBRkGAgAYBRkGAgAVGDgoKFk1CTUJNQk1CAUVZbWWAgAZCBUmAgkAOBUYFSkFBgAFJZUGAAUYBgIAFRWWAgAZCBUmAgkANgAWAAUVEBgVKQUGAAUlmQUJFQUJBWW2AgAVFRkFBZUICRUFB/bWUhAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABZYEABkIFSYCCQA39UaGVyZSBpcyBubyBoYW1zdGVyIHdpdGggdGhhdCBuYYFSYCCQA2AjgVKBYgAEvJBiAAInVltiAALekWIAArpWW4BgQAFRkFCQUJBWW1BQgpFQUGAAYABgAGAAGVlgIAGQgVJgIJADYABZkIFSgVJZYCABkIFSYCCQA2ABWZCBUoFSWWAgAZCBUmAgkANgAFmQgVKBUllgIAGQgVJgIJADYAOBUllgQAGQgVJgIJADYAFZkIFSgVJgIJADYQEsgVJgAGAAWvFZYCABkIFSYCCQA2ABgVJZYCABkIFSYCCQA2AAGVlgIAGQgVJgIJADYAAZWWAgAZCBUmAgkANgAFmQgVKBUllgIAGQgVJgIJADYAFZkIFSgVJZYCABkIFSYCCQA2AAWZCBUoFSWWAgAZCBUmAgkANgA4FSWWBAAZCBUmAgkANgAVmQgVKBUmAgkANgBoFSgVJZYCABkIFSYCCQA2AAWZCBUoFSWWAgAZCBUmAgkANgA4FSgVKQVltgIAFRUVlQgJFQUICAUZBgIAEgkFCQVltgIAFRUZBQWVCAkVBQYgACJ1aFMy4yLjDqJO+0
```

Maybe check which compiler are you using if you are facing the issue locally.

**Comment by AleksandarFilipov** [comment]
@mradkov yep it's true that it compiles in the testnet online compiler...

**Comment by nikita-fuchs** [comment]
@AleksandarFilipov so it does here, too: http://fireeditor.nikitafuchs.de/?highlight=5-45-3-1&contract=254d8ec6-c ;)

**Comment by maptuhec** [comment]
@nikita-fuchs @thepiwo  Can we merge this one?

File: examples/CryptoHamster/contracts/crypto-hamsters.aes [hunk 1]
```diff
@@ -38,11 +38,12 @@ contract CryptoHamster =
         put(state{index = (state.index + 1)})
 
     private function generate_random_dna(name: string) : int =
-        Chain.block_hash(Chain.block_height - 1) - Chain.timestamp + state.index
+        get_block_hash_bytes_as_int() - Chain.timestamp + state.index
 
-    public function test(name: string) : hash =
-        String.sha3(name)
+    private function get_block_hash_bytes_as_int() : int =
```
Comment (mradkov) [inline examples/CryptoHamster/contracts/crypto-hamsters.aes]: No need for `private` keyword anymore.

File: examples/CryptoHamster/contracts/crypto-hamsters.aes [hunk 2]
```diff
@@ -38,11 +38,12 @@ contract CryptoHamster =
         put(state{index = (state.index + 1)})
 
     private function generate_random_dna(name: string) : int =
```
Comment (mradkov) [inline examples/CryptoHamster/contracts/crypto-hamsters.aes]: No need for `private` keyword anymore.

File: examples/CryptoHamster/contracts/crypto-hamsters.aes [hunk 3]
```diff
@@ -9,26 +9,26 @@ contract CryptoHamster =
...
-    public function get_hamster_dna(name: string) : int =
+    entrypoint get_hamster_dna(name: string) : int =
         require(name_exists(name), "There is no hamster with that name!")
 
         let needed_hamster : hamster = state.map_hamsters[name]
 
         needed_hamster.dna
 
-    private function create_hamster_by_name_dna(name: string, dna: int) =
+    private stateful function create_hamster_by_name_dna(name: string, dna: int) =
```
Comment (mradkov) [inline examples/CryptoHamster/contracts/crypto-hamsters.aes]: No need for `private` keyword anymore.

**Pull request merged**